include/stdexec/execution.hpp

```diff
@@ -3292,91 +3403,72 @@ namespace stdexec {
+        // any point after the callback is registered. Beware. We are guaranteed, however, that
+        // __on_stop_request will not complete the operation or decrement the shared state's ref
+        // count until after __self has been added to the waiters list.
+        __self.__register_stop_callback();
+
+        // We haven't put __self in the waiters list yet and we are holding a ref count to
+        // __sh_state_, so nothing can happen to the __sh_state_ here.
+
+        // If the receiver's stop token has already been signaled, don't bother trying to start
+        // the underlying operation.
+        bool const __stop_requested =
+          stdexec::get_stop_token(stdexec::get_env(__rcvr)).stop_requested();
+        if (!__stop_requested) {
+          __self.__sh_state_->__try_start();
         }
 
-        // With the split algorithm, multiple split senders can be started simultaneously,
-        // but only one should start the async operation. The following loop atomically
-        // (re)tries to set the pointer to the head of the list to __state. When it finally
-        // succeeds, the prior value is checked. If it is nullptr, then this split
-        // sender has won the race and has the honor of starting the async operation.
-        do {
-          if (__old == __completion_state) {
-            __state.template __action<_Tag>(&__state, __action_kind::__notify);
-            return;
-          }
-          __state.__next_ = static_cast<__local_state_base*>(__old);
-        } while (!__head.compare_exchange_weak(
-          __old,
-          static_cast<void*>(&__state),
-#if STDEXEC_TSAN() && STDEXEC_GCC() && (__GNUC__ <= 11)
-// GCC-11 TSAN has a false positive with memory_order_release,
-// even though release ordering is correct.
-          std::memory_order_acq_rel,
-#else
-          std::memory_order_release,
-#endif
-          std::memory_order_acquire));
-
-        if constexpr (same_as<_Tag, __split::__split_t>) {
-          if (__old == nullptr) {
-            __shared_state->__start_op();
+        if (std::unique_lock __lock{__self.__sh_state_->__mutex_}) {
+          if (__self.__sh_state_->__waiters_.front() == __local_state_base::__get_tombstone()) {
+            // The operation has already completed. Instead of adding this operation to the list
+            // of waiters, pass the result to the receiver by calling __notify on the local state.
+            __lock.unlock();
+            __self.template __notify<_Tag>(&__self);
+          } else if (__stop_requested) {
+            __lock.unlock();
+            __self.__on_stop_.reset();
+            __self.__release_shared_state();
+            stdexec::set_stopped(static_cast<_Receiver&&>(__rcvr));
+          } else {
+            // The operation is still running. Add this operation to the waiters list.
+            __self.__sh_state_->__waiters_.push_front(&__self);
           }
         }
```

Comment:
This whole block of code seems like it should be a method on the `__shared_state` type.

e.g. given:
```c++
template<typename StopToken>
bool __shared_state::__try_add_waiter(__local_state_base* __waiter, StopToken st) noexcept {
  std::unique_lock __lock{__mutex_};
  if (__waiters_.front() == __local_state_base::__get_tombstone()) {
    __lock.unlock();
    __waiter->__notify(__waiter);
    return true;
  } else if (st.stop_requested()) {
    return false;
  } else {
    __waiters_.push_front(__waiter);
    return true;
  }
}
```

then this logic becomes:
```c++
const auto __st = stdexec::get_stop_token(stdexec::get_env(__rcvr));
if (!__st.stop_requested()) {
  __self.__sh_state->__try_start();
  if (__self.__sh_state->__try_add_waiter(__self, __st)) {
    // successfully added the waiter
    return;
  }
}

// Otherwise, failed to add the waiter because of a stop-request.
// Complete synchronously with set_stopped().
__self.__on_stop_.reset();
__self.__release_shared_state();
stdexec::set_stopped(static_cast<_Receiver&&>(__rcvr));
```

Reply:
nice